test/e2e/handler/dns_test.go

```diff
@@ -0,0 +1,102 @@
+package handler
+
+import (
+	"fmt"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	nmstate "github.com/nmstate/kubernetes-nmstate/api/shared"
+)
+
+func dnsConfig(searchDomain1, searchDomain2, upStream1, upStream2, ipFamily, dnsTestNic string) nmstate.State {
+	return nmstate.NewState(fmt.Sprintf(`dns-resolver:
+  config:
+   search:
+   - %s
+   - %s
+   server:
+   - %s
+   - %s
+interfaces:
+- name: %s
+  type: ethernet
+  state: up
+  %s:
+    auto-dns: false
+    enabled: true
+    dhcp: true
+
+`, searchDomain1, searchDomain2, upStream1, upStream2, dnsTestNic, ipFamily))
+}
+
+func dnsAbsent(dnsTestNic, ipFamily string) nmstate.State {
+	return nmstate.NewState(fmt.Sprintf(`dns-resolver:
+  config:
+    state: absent
+interfaces:
+- name: %s
+  type: ethernet
+  state: up
+  %s:
+    auto-dns: true
+`, dnsTestNic, ipFamily))
+}
+
+var _ = FDescribe("Dns configuration", func() {
+	Context("when desiredState is configured", func() {
+		var (
+			searchDomain1 = "fufu.ostest.test.metalkube.org"
+			searchDomain2 = "sometest.com"
+			upStream1V4   = "8.8.9.9"
+			upStream1V6   = "2001:db8::1:2"
+		)
+
+		Context("with V4 upstream servers", func() {
+			BeforeEach(func() {
+				// read primary DNS server from one of the nodes
+				serverList := dnsResolverForNode(nodes[0], "dns-resolver.running.server")
+				updateDesiredStateAndWait(dnsConfig(searchDomain1, searchDomain2, serverList[0], upStream1V4, "ipv4", dnsTestNic))
+			})
+			AfterEach(func() {
+				updateDesiredStateAndWait(dnsAbsent(dnsTestNic, "ipv4"))
+				for _, node := range nodes {
+					dnsResolverServerForNodeEventually(node).ShouldNot(ContainElement(upStream1V4))
+					dnsResolverSearchForNodeEventually(node).ShouldNot(ContainElement(searchDomain1))
+					dnsResolverSearchForNodeEventually(node).ShouldNot(ContainElement(searchDomain2))
+				}
+				resetDesiredStateForNodes()
+			})
+			It("should have the static V4 address", func() {
+				for _, node := range nodes {
+					dnsResolverServerForNodeEventually(node).Should(ContainElement(upStream1V4))
+					dnsResolverSearchForNodeEventually(node).Should(ContainElement(searchDomain1))
+					dnsResolverSearchForNodeEventually(node).Should(ContainElement(searchDomain2))
+				}
+			})
+		})
+		Context("with V6 upstream servers", func() {
+			BeforeEach(func() {
+				// read primary DNS server from one of the nodes
+				serverList := dnsResolverForNode(nodes[0], "dns-resolver.running.server")
+				updateDesiredStateAndWait(dnsConfig(searchDomain1, searchDomain2, serverList[0], upStream1V6, "ipv6", dnsTestNic))
```

Comment:
serverList[0] can be an IPv4 but for IPv4 you have auto-dns so it will fail since you cannot configure dns with auto-dns,

You need to pass all the list and disable auto-dns for both IPv4 and IPv6.